File: app/src/main/java/com/dhpcs/liquidity/ServerConnection.java
package com.dhpcs.liquidity;

import android.content.Context;
import android.os.Handler;
import android.os.HandlerThread;
import android.os.Looper;
import android.util.SparseArray;

import com.dhpcs.jsonrpc.JsonRpcMessage;
import com.dhpcs.jsonrpc.JsonRpcMessage$;
import com.dhpcs.jsonrpc.JsonRpcNotificationMessage;
import com.dhpcs.jsonrpc.JsonRpcRequestMessage;
import com.dhpcs.jsonrpc.JsonRpcResponseError;
import com.dhpcs.jsonrpc.JsonRpcResponseMessage;
import com.dhpcs.liquidity.models.Command;
import com.dhpcs.liquidity.models.Command$;
import com.dhpcs.liquidity.models.ErrorResponse;
import com.dhpcs.liquidity.models.Notification;
import com.dhpcs.liquidity.models.Notification$;
import com.dhpcs.liquidity.models.Response;
import com.dhpcs.liquidity.models.Response$;
import com.dhpcs.liquidity.models.ResultResponse;
import com.fasterxml.jackson.core.JsonParseException;
import com.google.android.gms.common.GooglePlayServicesNotAvailableException;
import com.google.android.gms.common.GooglePlayServicesRepairableException;
import com.google.android.gms.common.GooglePlayServicesUtil;
import com.google.android.gms.security.ProviderInstaller;
import com.squareup.okhttp.OkHttpClient;
import com.squareup.okhttp.ws.WebSocket;
import com.squareup.okhttp.ws.WebSocketCall;
import com.squareup.okhttp.ws.WebSocketListener;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.security.KeyManagementException;
import java.security.NoSuchAlgorithmException;

import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLSocketFactory;

import okio.Buffer;
import okio.BufferedSource;
import play.api.libs.json.JsResult;
import play.api.libs.json.Json;
import scala.Int;
import scala.Option;
import scala.util.Right;

public class ServerConnection implements WebSocketListener {

    // TODO: Can this just be boolean?
    // TODO: Check NetworkUtils.isNetworkAvailable somewhere
    public enum ConnectionState {
        CONNECTING,
        CONNECTED,
        DISCONNECTING,
        DISCONNECTED
    }

    public interface ConnectionStateListener {

        void onStateChanged(ConnectionState connectionState);

    }

    public interface NotificationListener {

        void onNotificationReceived(Notification notification);

    }

    public static abstract class ResponseCallback {

        void onErrorReceived(ErrorResponse errorResponse) {
            throw new RuntimeException(errorResponse.toString());
        }

        abstract void onResultReceived(ResultResponse resultResponse);

        // TODO: Timeouts?

    }

    private static final String SERVER_ENDPOINT = "https://liquidity.dhpcs.com/ws";
    // TODO
    private static final long PING_PERIOD = 5000;

    private static SSLSocketFactory getSslSocketFactory(Context context) {
        try {
            SSLContext sslContext = SSLContext.getInstance("TLS");
            try {
                sslContext.init(
                        ClientKey.getInstance(context).getKeyManagers(),
                        ServerTrust.getInstance(context).getTrustManagers(),
                        null
                );
            } catch (KeyManagementException e) {
                throw new Error(e);
            }
            return sslContext.getSocketFactory();
        } catch (NoSuchAlgorithmException e) {
            throw new Error(e);
        }
    }

    private final Logger log = LoggerFactory.getLogger(getClass());

    private final Runnable pingRunnable = new Runnable() {

        @Override
        public void run() {
            try {
                webSocket.sendPing(null);
                connectionHandler.postDelayed(this, PING_PERIOD);
            } catch (IOException e) {
                log.warn(e.getMessage(), e);
            }
        }

    };

    private final Handler defaultHandler;
    private final ConnectionStateListener connectionStateListener;
    private final Handler connectionStateHandler;
    private final NotificationListener notificationListener;
    private final Handler notificationHandler;
    private final OkHttpClient client;

    private Handler connectionHandler;
    private WebSocketCall webSocketCall;
    private WebSocket webSocket;

    private final SparseArray<JsonRpcRequestMessage> requests = new SparseArray<>();
    private final SparseArray<ResponseCallback> requestResponseCallbacks = new SparseArray<>();
    private final SparseArray<Handler> requestResponseHandlers = new SparseArray<>();

    private int lastCommandIdentifier;

    public ServerConnection(Context context,
                            ConnectionStateListener connectionStateListener,
                            Handler connectionStateHandler,
                            NotificationListener notificationListener,
                            Handler notificationHandler) {

        try {
            ProviderInstaller.installIfNeeded(context);
        } catch (GooglePlayServicesRepairableException e) {
            log.warn("GooglePlayServicesRepairableException", e);
            GooglePlayServicesUtil.showErrorNotification(
                    e.getConnectionStatusCode(),
                    context
            );
        } catch (GooglePlayServicesNotAvailableException e) {
            log.warn("GooglePlayServicesNotAvailableException", e);
        }

        this.defaultHandler = new Handler(Looper.getMainLooper());
        this.connectionStateListener = connectionStateListener;
        this.connectionStateHandler = connectionStateHandler == null ?
                defaultHandler : connectionStateHandler;
        this.notificationListener = notificationListener;
        this.notificationHandler = notificationHandler == null ?
                defaultHandler : notificationHandler;
        this.client = new OkHttpClient();
        this.client.setHostnameVerifier(ServerTrust.getInstance(context).getHostnameVerifier());
        this.client.setSslSocketFactory(getSslSocketFactory(context));
    }

    public ServerConnection(Context context,
                            ConnectionStateListener connectionStateListener,
                            NotificationListener notificationListener) {
        this(context,
                connectionStateListener,
                null,
                notificationListener,
                null);
    }

    public void connect() {
        log.debug("Creating websocket call");
        if (connectionHandler != null) {
            throw new IllegalStateException("Already connecting/connected");
        }
        HandlerThread handlerThread = new HandlerThread("Connection");
        handlerThread.start();
        connectionHandler = new Handler(handlerThread.getLooper());
        connectionHandler.post(new Runnable() {

            @Override
            public void run() {
                connectionStateHandler.post(new Runnable() {

                    @Override
                    public void run() {
                        connectionStateListener.onStateChanged(ConnectionState.CONNECTING);
                    }

                });
                webSocketCall = WebSocketCall.create(
                        client,
                        new com.squareup.okhttp.Request.Builder()
                                .url(SERVER_ENDPOINT)
                                .build()
                );
                webSocketCall.enqueue(ServerConnection.this);
            }

        });
    }

    public void disconnect() {
        disconnect(1000, "Bye");
    }

    private void disconnect(final int code, final String reason) {
        if (connectionHandler == null) {
            throw new IllegalStateException("Not connecting/connected");
        }
        connectionHandler.post(new Runnable() {

            @Override
            public void run() {
                connectionStateHandler.post(new Runnable() {

                    @Override
                    public void run() {
                        connectionStateListener.onStateChanged(ConnectionState.DISCONNECTING);
                    }

                });
                if (webSocket == null) {
                    log.debug("Cancelling websocket call");
                    webSocketCall.cancel();
                    webSocketCall = null;
                } else {
                    try {
                        log.debug("Closing websocket");
                        webSocket.close(code, reason);
                        webSocket = null;
                    } catch (IOException e) {
                        log.warn(e.getMessage(), e);
                    }
                }
            }

        });
    }

    @Override
    public void onClose(int code, String reason) {
        log.debug("WebSocket closed. Code: {}, reason: {}", code, reason);
        connectionHandler.post(new Runnable() {

            @Override
            public void run() {
                connectionHandler.removeCallbacks(pingRunnable);
                connectionStateHandler.post(new Runnable() {

                    @Override
                    public void run() {
                        connectionStateListener.onStateChanged(ConnectionState.DISCONNECTED);
                    }

                });
                webSocketCall = null;
                webSocket = null;
                connectionHandler.getLooper().quit();
                connectionHandler = null;
            }

        });
    }

    @Override
    public void onFailure(IOException e) {
        log.warn(e.getMessage(), e);
    }

    @Override
    public void onMessage(BufferedSource payload, WebSocket.PayloadType type) throws IOException {
        switch (type) {
            case TEXT:

                try {

                    JsResult<JsonRpcMessage> jsonRpcMessageJsResult = Json.fromJson(
                            Json.parse(payload.readUtf8()),
                            JsonRpcMessage$.MODULE$.JsonRpcMessageFormat()
                    );

                    if (jsonRpcMessageJsResult.isError()) {

                        log.warn("Invalid JSON-RPC message: {}", jsonRpcMessageJsResult);

                    } else {

                        JsonRpcMessage jsonRpcMessage = jsonRpcMessageJsResult.get();

                        if (jsonRpcMessage instanceof JsonRpcResponseMessage) {

                            final JsonRpcResponseMessage jsonRpcResponseMessage
                                    = (JsonRpcResponseMessage) jsonRpcMessage;

                            log.debug("Received JSON-RPC response message: {}",
                                    jsonRpcResponseMessage);

                            if (jsonRpcResponseMessage.id().isEmpty()) {

                                JsonRpcResponseError jsonRpcResponseError =
                                        jsonRpcResponseMessage.eitherErrorOrResult().left().get();

                                log.warn("Received JSON-RPC response error: {}",
                                        jsonRpcResponseError);

                            } else {

                                connectionHandler.post(new Runnable() {

                                    @Override
                                    public void run() {

                                        int commandIdentifier = Int.unbox(
                                                (jsonRpcResponseMessage.id().get().right().get())
                                        );
                                        JsonRpcRequestMessage jsonRpcRequestMessage =
                                                requests.get(commandIdentifier);
                                        requests.remove(commandIdentifier);
                                        final ResponseCallback responseCallback =
                                                requestResponseCallbacks.get(commandIdentifier);
                                        requestResponseCallbacks.remove(commandIdentifier);
                                        final Handler responseHandler =
                                                requestResponseHandlers.get(commandIdentifier);
                                        requestResponseHandlers.remove(commandIdentifier);

                                        JsResult<Response> responseJsResult =
                                                Response$.MODULE$.read(
                                                        jsonRpcResponseMessage,
                                                        jsonRpcRequestMessage.method()
                                                );

                                        if (responseJsResult.isError()) {

                                            log.warn("Invalid Response: {}", responseJsResult);

                                        } else {

                                            final Response response = responseJsResult.get();

                                            if (response instanceof ErrorResponse) {

                                                log.warn("Received response error: {}", response);

                                                responseHandler.post(new Runnable() {

                                                    @Override
                                                    public void run() {
                                                        responseCallback.onErrorReceived(
                                                                (ErrorResponse) response
                                                        );
                                                    }

                                                });

                                            } else {

                                                responseHandler.post(new Runnable() {

                                                    @Override
                                                    public void run() {
                                                        responseCallback.onResultReceived(
                                                                (ResultResponse) response
                                                        );
                                                    }

                                                });

                                            }

                                        }
                                    }

                                });

                            }

                        } else if (jsonRpcMessage instanceof JsonRpcNotificationMessage) {

                            JsonRpcNotificationMessage jsonRpcNotificationMessage =
                                    (JsonRpcNotificationMessage) jsonRpcMessage;

                            log.debug("Received JSON-RPC notification message: {}",
                                    jsonRpcNotificationMessage);

                            Option<JsResult<Notification>> maybeJsResultNotification =
                                    Notification$.MODULE$.read(
                                            jsonRpcNotificationMessage
                                    );

                            if (maybeJsResultNotification.isEmpty()) {

                                log.warn("Received notification for unimplemented method");

                            } else {

                                final JsResult<Notification> jsResultNotification =
                                        maybeJsResultNotification.get();

                                if (jsResultNotification.isError()) {

                                    log.warn("Invalid Notification: {}", jsResultNotification);

                                } else {

                                    connectionHandler.post(new Runnable() {

                                        @Override
                                        public void run() {

                                            notificationHandler.post(new Runnable() {

                                                @Override
                                                public void run() {
                                                    notificationListener.onNotificationReceived(
                                                            jsResultNotification.get()
                                                    );
                                                }

                                            });

                                        }

                                    });

                                };

                            }

                        }

                    }

                } catch (JsonParseException e) {
                    log.warn("Invalid JSON: {}", e);
                }

                break;
            case BINARY:

               /*
                * See https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent.
                */
                disconnect(1003, "I don't support binary data");

                break;
            default:
                throw new IllegalStateException("Unknown payload type: " + type);
        }
        payload.close();
    }

    @Override
    public void onOpen(final WebSocket webSocket,
                       com.squareup.okhttp.Request request,
                       com.squareup.okhttp.Response response) throws IOException {
        log.debug("WebSocket opened");
        connectionHandler.post(new Runnable() {

            @Override
            public void run() {
                ServerConnection.this.webSocketCall = null;
                ServerConnection.this.webSocket = webSocket;
                connectionStateHandler.post(new Runnable() {

                    @Override
                    public void run() {
                        connectionStateListener.onStateChanged(ConnectionState.CONNECTED);
                    }

                });
                connectionHandler.postDelayed(pingRunnable, PING_PERIOD);
            }

        });
    }

    @Override
    public void onPong(Buffer payload) {
        String payloadString = null;
        if (payload != null) {
            payloadString = payload.readUtf8();
            payload.close();
        }
        log.debug("Received pong: {}", payloadString);
    }

    public void sendCommand(final Command command,
                            final ResponseCallback responseCallback,
                            final Handler responseHandler) {
        log.debug("Sending command: {}", command);
        connectionHandler.post(new Runnable() {

            @Override
            public void run() {
                if (webSocket == null) {
                    throw new IllegalStateException("Not connected");
                }
                try {
                    final int commandIdentifier = lastCommandIdentifier++;
                    final JsonRpcRequestMessage jsonRpcRequestMessage = Command$.MODULE$.write(
                            command,
                            Right.<String, Object>apply(Int.box(commandIdentifier))
                    );
                    webSocket.sendMessage(
                            WebSocket.PayloadType.TEXT,
                            new Buffer().writeUtf8(
                                    Json.stringify(
                                            Json.toJson(
                                                    jsonRpcRequestMessage,
                                                    JsonRpcRequestMessage
                                                            .JsonRpcRequestMessageFormat()
                                            )
                                    )
                            )
                    );
                    requests.append(commandIdentifier, jsonRpcRequestMessage);
                    requestResponseCallbacks.append(commandIdentifier, responseCallback);
                    requestResponseHandlers.append(commandIdentifier, responseHandler);
                } catch (IOException e) {
                    log.warn(e.getMessage(), e);
                }
            }

        });
    }

    public void sendCommand(final Command command,
                            final ResponseCallback responseCallback) {
        sendCommand(
                command,
                responseCallback,
                defaultHandler);
    }

}

File: app/src/main/scala/com/dhpcs/liquidity/ServerConnection.scala
package com.dhpcs.liquidity

import java.io.{File, IOException}
import java.util.concurrent.TimeUnit
import javax.net.ssl.{KeyManager, SSLContext, SSLException, TrustManager}

import android.content.{BroadcastReceiver, Context, Intent, IntentFilter}
import android.net.ConnectivityManager
import android.os.{Handler, HandlerThread, Looper}
import com.dhpcs.jsonrpc._
import com.dhpcs.liquidity.ServerConnection._
import com.dhpcs.liquidity.models._
import okhttp3.ws.{WebSocket, WebSocketCall, WebSocketListener}
import okhttp3.{OkHttpClient, RequestBody, ResponseBody}
import okio.Buffer
import play.api.libs.json.Json

import scala.util.{Failure, Right, Success, Try}

object ServerConnection {

  PRNGFixes.apply()

  sealed trait ConnectionState

  case object UNAVAILABLE extends ConnectionState

  case object GENERAL_FAILURE extends ConnectionState

  case object TLS_ERROR extends ConnectionState

  case object UNSUPPORTED_VERSION extends ConnectionState

  case object AVAILABLE extends ConnectionState

  case object CONNECTING extends ConnectionState

  case object WAITING_FOR_VERSION_CHECK extends ConnectionState

  case object ONLINE extends ConnectionState

  case object DISCONNECTING extends ConnectionState

  trait ConnectionStateListener {

    def onConnectionStateChanged(connectionState: ConnectionState)

  }

  trait NotificationReceiptListener {

    def onZoneNotificationReceived(zoneNotification: ZoneNotification)

  }

  class ConnectionRequestToken

  trait ResponseCallback {

    def onErrorReceived(errorResponse: ErrorResponse)

    def onResultReceived(resultResponse: ResultResponse) = ()

  }

  private sealed trait State

  private sealed trait IdleState extends State

  private case object UnavailableIdleState extends IdleState

  private case object GeneralFailureIdleState extends IdleState

  private case object TlsErrorIdleState extends IdleState

  private case object UnsupportedVersionIdleState extends IdleState

  private case object AvailableIdleState extends IdleState

  private case class ActiveState(handler: Handler) extends State {

    var subState: SubState = _

  }

  private sealed trait SubState

  private case class ConnectingSubState(webSocketCall: WebSocketCall) extends SubState

  private sealed trait ConnectedSubState extends SubState {

    val webSocket: WebSocket

  }

  private case class WaitingForVersionCheckSubState(webSocket: WebSocket) extends ConnectedSubState

  private case class OnlineSubState(webSocket: WebSocket) extends ConnectedSubState

  private case object DisconnectingSubState extends SubState

  private case class PendingRequest(requestMessage: JsonRpcRequestMessage,
                                    callback: ResponseCallback)

  private sealed trait CloseCause

  private case object GeneralFailure extends CloseCause

  private case object TlsError extends CloseCause

  private case object UnsupportedVersion extends CloseCause

  private case object ServerDisconnect extends CloseCause

  private case object ClientDisconnect extends CloseCause

  private val ServerEndpoint = "https://liquidity.dhpcs.com/ws"

  private var instance: ServerConnection = _

  private def asyncPost(handler: Handler)(body: => Unit) {
    handler.post(new Runnable() {

      override def run() = body

    })
  }

  private def createSslSocketFactory(keyManagers: Array[KeyManager],
                                     trustManagers: Array[TrustManager]) = {
    val sslContext = SSLContext.getInstance("TLS")
    sslContext.init(
      keyManagers,
      trustManagers,
      null
    )
    sslContext.getSocketFactory
  }

  def getInstance(context: Context,
                  filesDir: File,
                  clientId: String) = {
    if (instance == null) {
      instance = new ServerConnection(context, filesDir, clientId)
    }
    instance
  }

  private def readJsonRpcMessage(jsonString: String): Either[String, JsonRpcMessage] =

    Try(Json.parse(jsonString)) match {

      case Failure(exception) =>

        Left(s"Invalid JSON: $exception")

      case Success(json) =>

        Json.fromJson[JsonRpcMessage](json).fold({ errors =>
          Left(s"Invalid JSON-RPC message: $errors")
        }, Right(_))

    }

}

class ServerConnection private(context: Context,
                               filesDir: File,
                               clientId: String) extends WebSocketListener {

  private lazy val client = new OkHttpClient.Builder()
    .sslSocketFactory(createSslSocketFactory(
      ClientKey.getKeyManagers(filesDir, clientId),
      ServerTrust.getTrustManagers(context.getResources.openRawResource(R.raw.liquidity_dhpcs_com))
    ))
    .readTimeout(0, TimeUnit.SECONDS)
    .writeTimeout(0, TimeUnit.SECONDS)
    .build()

  private val mainHandler = new Handler(Looper.getMainLooper)

  private val connectionStateFilter = new IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION)
  private val connectionStateReceiver = new BroadcastReceiver {

    override def onReceive(context: Context, intent: Intent) = intent.getAction match {

      case ConnectivityManager.CONNECTIVITY_ACTION =>

        handleConnectivityStateChange()

      case unmatched =>

        sys.error(s"Received unexpected broadcast for action $unmatched")

    }

  }

  private var pendingRequests = Map.empty[BigDecimal, PendingRequest]
  private var commandIdentifier = BigDecimal(0)
  private var state: State = UnavailableIdleState
  private var hasFailed: Boolean = _

  private var _connectionState: ConnectionState = UNAVAILABLE

  private var connectionStateListeners = Set.empty[ConnectionStateListener]
  private var connectRequestTokens = Set.empty[ConnectionRequestToken]

  private var notificationReceiptListeners = Set.empty[NotificationReceiptListener]

  handleConnectivityStateChange()

  def clientKey = ClientKey.getPublicKey(filesDir, clientId)

  private def connect() = state match {

    case _: ActiveState =>

      sys.error("Already connecting/connected/disconnecting")

    case UnavailableIdleState =>

      sys.error("Connection unavailable")

    case AvailableIdleState
         | GeneralFailureIdleState
         | TlsErrorIdleState
         | UnsupportedVersionIdleState =>

      doOpen()

  }

  def connectionState = _connectionState

  private def disconnect(code: Int) = state match {

    case _: IdleState =>

      sys.error("Already disconnected")

    case activeState: ActiveState =>

      asyncPost(activeState.handler)(activeState.subState match {

        case DisconnectingSubState =>

          sys.error("Already disconnecting")

        case ConnectingSubState(webSocketCall) =>

          activeState.subState = DisconnectingSubState
          asyncPost(mainHandler) {
            _connectionState = DISCONNECTING
            connectionStateListeners.foreach(_.onConnectionStateChanged(_connectionState))
          }

          webSocketCall.cancel()

        case WaitingForVersionCheckSubState(webSocket) =>

          try {
            webSocket.close(code, null)
          } catch {
            case _: IOException =>
          }

        case OnlineSubState(webSocket) =>

          activeState.subState = DisconnectingSubState
          asyncPost(mainHandler) {
            _connectionState = DISCONNECTING
            connectionStateListeners.foreach(_.onConnectionStateChanged(_connectionState))
          }

          try {
            webSocket.close(code, null)
          } catch {
            case _: IOException =>
          }

      })

  }

  private def doClose(handler: Handler,
                      closeCause: CloseCause,
                      reconnect: Boolean = false) {
    handler.getLooper.quit()
    commandIdentifier = 0
    pendingRequests = Map.empty
    asyncPost(mainHandler) {
      closeCause match {

        case GeneralFailure =>

          hasFailed = true
          state = GeneralFailureIdleState
          _connectionState = GENERAL_FAILURE
          connectionStateListeners.foreach(_.onConnectionStateChanged(_connectionState))

        case TlsError =>

          hasFailed = true
          state = TlsErrorIdleState
          _connectionState = TLS_ERROR
          connectionStateListeners.foreach(_.onConnectionStateChanged(_connectionState))

        case UnsupportedVersion =>

          hasFailed = true
          state = UnsupportedVersionIdleState
          _connectionState = UNSUPPORTED_VERSION
          connectionStateListeners.foreach(_.onConnectionStateChanged(_connectionState))

        case ServerDisconnect =>

          hasFailed = true
          state = AvailableIdleState
          _connectionState = AVAILABLE
          connectionStateListeners.foreach(_.onConnectionStateChanged(_connectionState))

        case ClientDisconnect =>

          hasFailed = false
          if (connectRequestTokens.nonEmpty) {
            doOpen()
          } else {
            state = AvailableIdleState
            _connectionState = AVAILABLE
            connectionStateListeners.foreach(_.onConnectionStateChanged(_connectionState))
          }

      }
    }
  }

  private def doOpen() {
    val handlerThread = new HandlerThread("ServerConnection")
    handlerThread.start()
    val handler = new Handler(handlerThread.getLooper)
    val activeState = ActiveState(handler)
    state = activeState
    asyncPost(handler) {
      val webSocketCall = WebSocketCall.create(
        client,
        new okhttp3.Request.Builder().url(ServerEndpoint).build
      )
      webSocketCall.enqueue(this)
      activeState.subState = ConnectingSubState(webSocketCall)
    }
    _connectionState = CONNECTING
    connectionStateListeners.foreach(_.onConnectionStateChanged(_connectionState))
  }

  private def handleConnectivityStateChange() =
    if (!isConnectionAvailable) {
      state match {

        case AvailableIdleState
             | GeneralFailureIdleState
             | TlsErrorIdleState
             | UnsupportedVersionIdleState =>

          state = UnavailableIdleState
          _connectionState = UNAVAILABLE
          connectionStateListeners.foreach(_.onConnectionStateChanged(_connectionState))

        case _ =>

      }
    } else {
      state match {

        case UnavailableIdleState =>

          state = AvailableIdleState
          _connectionState = AVAILABLE
          connectionStateListeners.foreach(_.onConnectionStateChanged(_connectionState))

        case _ =>

      }
    }

  def isConnectionAvailable = {
    val activeNetwork = context
      .getSystemService(Context.CONNECTIVITY_SERVICE)
      .asInstanceOf[ConnectivityManager]
      .getActiveNetworkInfo
    activeNetwork != null && activeNetwork.isConnected
  }

  override def onClose(code: Int, reason: String) =
    asyncPost(mainHandler)(state match {

      case _: IdleState =>

        sys.error("Already disconnected")

      case activeState: ActiveState =>

        asyncPost(activeState.handler)(activeState.subState match {

          case _: ConnectingSubState =>

            sys.error("Not connected or disconnecting")

          case _: WaitingForVersionCheckSubState =>

            doClose(
              activeState.handler,
              UnsupportedVersion
            )

          case _: OnlineSubState =>

            doClose(
              activeState.handler,
              ServerDisconnect
            )

          case DisconnectingSubState =>

            doClose(
              activeState.handler,
              ClientDisconnect
            )

        })

    })

  override def onFailure(e: IOException, response: okhttp3.Response) =
    asyncPost(mainHandler)(state match {

      case _: IdleState =>

        sys.error("Already disconnected")

      case activeState: ActiveState =>

        asyncPost(activeState.handler)(activeState.subState match {

          case DisconnectingSubState =>

            doClose(
              activeState.handler,
              ClientDisconnect
            )

          case _ =>

            if (response == null) {

              e match {

                case _: SSLException =>

                  /*
                   * Client rejected server certificate.
                   */
                  doClose(
                    activeState.handler,
                    TlsError
                  )

                case _ =>

                  doClose(
                    activeState.handler,
                    GeneralFailure
                  )

              }

            } else {

              if (response.code == 400) {

                /*
                 * Server rejected client certificate.
                 */
                doClose(
                  activeState.handler,
                  TlsError
                )

              } else {

                doClose(
                  activeState.handler,
                  GeneralFailure
                )

              }

            }

        })

    })

  override def onMessage(message: ResponseBody) = message.contentType match {

    case WebSocket.BINARY =>

      sys.error("Received binary frame")

    case WebSocket.TEXT =>

      readJsonRpcMessage(message.string) match {

        case Left(error) =>

          sys.error(error)

        case Right(jsonRpcMessage) =>

          asyncPost(mainHandler)(state match {

            case _: IdleState =>

              sys.error("Not connected")

            case activeState: ActiveState =>

              jsonRpcMessage match {

                case jsonRpcNotificationMessage: JsonRpcNotificationMessage =>

                  asyncPost(activeState.handler) {

                    Notification.read(jsonRpcNotificationMessage).fold {
                      sys.error(s"No notification type exists with method" +
                        s"=${jsonRpcNotificationMessage.method}")
                    }(_.fold({ errors =>
                      sys.error(s"Invalid Notification: $errors")
                    }, {

                      case SupportedVersionsNotification(compatibleVersionNumbers) =>

                        activeState.subState match {

                          case _: ConnectingSubState =>

                            sys.error("Not connected")

                          case _: OnlineSubState =>

                            sys.error("Already online")

                          case WaitingForVersionCheckSubState(webSocket) =>

                            if (!compatibleVersionNumbers.contains(VersionNumber)) {

                              asyncPost(mainHandler)(
                                disconnect(1001)
                              )

                            } else {

                              asyncPost(activeState.handler) {
                                activeState.subState = OnlineSubState(webSocket)
                                asyncPost(mainHandler) {
                                  _connectionState = ONLINE
                                  connectionStateListeners.foreach(
                                    _.onConnectionStateChanged(_connectionState)
                                  )
                                }
                              }

                            }

                          case DisconnectingSubState =>

                        }

                      case KeepAliveNotification =>

                        activeState.subState match {

                          case _: ConnectingSubState =>

                            sys.error("Not connected")

                          case _: WaitingForVersionCheckSubState =>

                            sys.error("Waiting for version check")

                          case _: OnlineSubState =>

                          case DisconnectingSubState =>

                        }

                      case zoneNotification: ZoneNotification =>

                        activeState.subState match {

                          case _: ConnectingSubState =>

                            sys.error("Not connected")

                          case _: WaitingForVersionCheckSubState =>

                            sys.error("Waiting for version check")

                          case _: OnlineSubState =>

                            asyncPost(activeState.handler)(
                              asyncPost(mainHandler)(
                                notificationReceiptListeners.foreach(
                                  _.onZoneNotificationReceived(zoneNotification)
                                )
                              )
                            )

                          case DisconnectingSubState =>

                        }

                    }))

                  }

                case jsonRpcResponseMessage: JsonRpcResponseMessage =>

                  asyncPost(activeState.handler)(activeState.subState match {

                    case _: ConnectingSubState =>

                      sys.error("Not connected")

                    case _: WaitingForVersionCheckSubState =>

                      sys.error("Waiting for version check")

                    case _: OnlineSubState =>

                      jsonRpcResponseMessage.id.fold {
                        sys.error(s"JSON-RPC message ID missing, jsonRpcResponseMessage" +
                          s".eitherErrorOrResult=${jsonRpcResponseMessage.eitherErrorOrResult}")
                      } { id =>
                        id.right.toOption.fold {
                          sys.error(s"JSON-RPC message ID was not a number, id=$id")
                        } { commandIdentifier =>
                          asyncPost(activeState.handler)(
                            pendingRequests.get(commandIdentifier).fold {
                              sys.error(s"No pending request exists with commandIdentifier" +
                                s"=$commandIdentifier")
                            } { pendingRequest =>
                              pendingRequests = pendingRequests - commandIdentifier
                              Response.read(
                                jsonRpcResponseMessage,
                                pendingRequest.requestMessage.method
                              ).fold({ errors =>
                                sys.error(s"Invalid Response: $errors")
                              }, {

                                case Left(errorResponse) =>

                                  asyncPost(mainHandler)(
                                    pendingRequest.callback.onErrorReceived(errorResponse)
                                  )

                                case Right(resultResponse) =>

                                  asyncPost(mainHandler)(
                                    pendingRequest.callback.onResultReceived(resultResponse)
                                  )

                              })
                            }
                          )
                        }
                      }

                    case DisconnectingSubState =>

                  })

                case jsonRpc_Message =>

                  asyncPost(activeState.handler)(activeState.subState match {

                    case _: ConnectingSubState =>

                      sys.error("Not connected")

                    case _: WaitingForVersionCheckSubState =>

                      sys.error("Waiting for version check")

                    case _: OnlineSubState =>

                      sys.error(s"Received $jsonRpc_Message")

                    case DisconnectingSubState =>

                  })

              }

          })

      }

  }

  override def onOpen(webSocket: WebSocket, response: okhttp3.Response) =
    asyncPost(mainHandler)(state match {

      case _: IdleState =>

        sys.error("Not connecting")

      case activeState: ActiveState =>

        asyncPost(activeState.handler)(activeState.subState match {

          case _: ConnectedSubState | DisconnectingSubState =>

            sys.error("Not connecting")

          case _: ConnectingSubState =>

            activeState.subState = WaitingForVersionCheckSubState(webSocket)
            asyncPost(mainHandler) {
              _connectionState = WAITING_FOR_VERSION_CHECK
              connectionStateListeners.foreach(_.onConnectionStateChanged(_connectionState))
            }

        })

    })

  override def onPong(payload: Buffer) = ()

  def registerListener(listener: ConnectionStateListener) =
    if (!connectionStateListeners.contains(listener)) {
      if (connectionStateListeners.isEmpty) {
        context.registerReceiver(connectionStateReceiver, connectionStateFilter)
        handleConnectivityStateChange()
      }
      connectionStateListeners = connectionStateListeners + listener
      listener.onConnectionStateChanged(_connectionState)
    }

  def registerListener(listener: NotificationReceiptListener) =
    if (!notificationReceiptListeners.contains(listener)) {
      notificationReceiptListeners = notificationReceiptListeners + listener
    }

  def requestConnection(token: ConnectionRequestToken, retry: Boolean) {
    if (!connectRequestTokens.contains(token)) {
      connectRequestTokens = connectRequestTokens + token
    }
    if ((_connectionState == ServerConnection.AVAILABLE
      || _connectionState == ServerConnection.GENERAL_FAILURE
      || _connectionState == ServerConnection.TLS_ERROR
      || _connectionState == ServerConnection.UNSUPPORTED_VERSION)
      && (!hasFailed || retry)) {

      connect()

    }
  }

  def sendCommand(command: Command, responseCallback: ResponseCallback) = state match {

    case _: IdleState =>

      sys.error("Not connected")

    case activeState: ActiveState =>

      asyncPost(activeState.handler)(activeState.subState match {

        case _: ConnectingSubState | DisconnectingSubState =>

          sys.error(s"Not connected")

        case _: WaitingForVersionCheckSubState =>

          sys.error("Waiting for version check")

        case onlineSubState: OnlineSubState =>

          val jsonRpcRequestMessage = Command.write(command, Some(Right(commandIdentifier)))
          commandIdentifier = commandIdentifier + 1
          try {
            onlineSubState.webSocket.sendMessage(
              RequestBody.create(
                WebSocket.TEXT,
                Json.stringify(
                  Json.toJson(jsonRpcRequestMessage)
                )
              )
            )
            pendingRequests = pendingRequests +
              (jsonRpcRequestMessage.id.get.right.get ->
                PendingRequest(jsonRpcRequestMessage, responseCallback))
          } catch {

            /*
             * We do nothing here because we count on receiving a call to onFailure due to a
             * matching read error.
             */
            case _: IOException =>
          }

      })

  }

  def unregisterListener(listener: ConnectionStateListener) =
    if (connectionStateListeners.contains(listener)) {
      connectionStateListeners = connectionStateListeners - listener
      if (connectionStateListeners.isEmpty) {
        context.unregisterReceiver(connectionStateReceiver)
      }
    }

  def unregisterListener(listener: NotificationReceiptListener) =
    if (notificationReceiptListeners.contains(listener)) {
      notificationReceiptListeners = notificationReceiptListeners - listener
    }

  def unrequestConnection(token: ConnectionRequestToken) =
    if (connectRequestTokens.contains(token)) {
      connectRequestTokens = connectRequestTokens - token
      if (connectRequestTokens.isEmpty) {
        if (_connectionState == ServerConnection.CONNECTING
          || _connectionState == ServerConnection.WAITING_FOR_VERSION_CHECK
          || _connectionState == ServerConnection.ONLINE) {

          disconnect(1001)

        }
      }
    }

}
